### PR TITLE
Refactoring dotnet-compile and adding a new RuntimeOutputDir to compile scripts variables.

### PR DESCRIPTION
**Pull request by livarcocc**
Refactoring dotnet-compile to make adding the script variable testable. So far, moved it to have a compiler controller and two separate compilers: native and managed. Also moved the script runner to its own class so that we can mock it into the managed controller.

Adding a ICommand interface and factory that we can use to mock Commands and avoid actually running them in unit tests.

Adding a new RuntimeOutputDir variable to the compile scripts.

### CODE REVIEW

**Comment by livarcocc** [comment]
cc @piotrpMSFT @brthor @Sridhar-MS

File: src/Microsoft.DotNet.Cli.Utils/ICommandFactory.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using Microsoft.DotNet.Cli.Utils;
+using NuGet.Frameworks;
+
+namespace Microsoft.DotNet.Tools.Compiler
```
Comment (piotrpMSFT) [inline src/Microsoft.DotNet.Cli.Utils/ICommandFactory.cs]: Why does this live in the Compiler namespace?

File: src/Microsoft.DotNet.Cli.Utils/DotNetCommandFactory.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using Microsoft.DotNet.Cli.Utils;
+using NuGet.Frameworks;
+
+namespace Microsoft.DotNet.Tools.Compiler
+{
```
Comment (piotrpMSFT) [inline src/Microsoft.DotNet.Cli.Utils/DotNetCommandFactory.cs]: and here

**Comment by piotrpMSFT** [comment]
:shipit:

**Comment by livarcocc** [comment]
@dotnet-bot Test OSX Release Build please

**Comment by livarcocc** [comment]
@dotnet-bot Test Windows_NT Debug Build please
@dotnet-bot Test Windows_NT Release Build please

**Pull request merged**